### ISSUE TITLE
Switch to Proof of Interaction, an algorithm that requires small amount of computational power while being secure. [feat]

### ISSUE THREAD

**Issue by JushBJJ** [dogecoin/dogecoin#2059]
# Feature Request

Full paper here: https://arxiv.org/abs/2002.07763

Proof of Interaction requires a small amounts of computational power that is resistant to selfish mining. In addition, it solves the issues of Proof of Work's computational power and Proof of Stake's security issues.

The way it works is that...

"when a user wants to access a resource in a server that is heavily requested, the server can ask the user to perform a tour of a given length in the network. This tour consists of accessing randomly a list of nodes, own by the same provider as the server. After the tour, a user can prove to the server that it has completed the task and can then retrieve the resource."

![image](https://user-images.githubusercontent.com/36951064/119435135-e9146180-bd5c-11eb-9038-a526813aa877.png)

**Comment by simonmadsen** [comment]
Something like this, or PoS (Proof of state) might also be a viable route.
I stake eth but won't mind to stake DOGE on the same machine which, because of PoS is pretty lightweight, should be no problem.

**Comment by JushBJJ** [comment]
> Something like this, or PoS (Proof of state) might also be a viable route.
> I stake eth but won't mind to stake DOGE on the same machine which, because of PoS is pretty lightweight, should be no problem.

Proof of Interaction solves the problems Proof of Stake has.

**Comment by simonmadsen** [comment]
there must be more too it?
or the eth devs had also chosen PoI,
actually maybe they will have PoI also, i think they call it stateless?

________________________________

**Comment by JushBJJ** [comment]
> there must be more too it?
> or the eth devs had also chosen PoI,
> actually maybe they will have PoI also, i think they call it stateless?
Could you rephrase that i couldn't understand

**Comment by simonmadsen** [comment]
> > there must be more too it?
> > or the eth devs had also chosen PoI,
> > actually maybe they will have PoI also, i think they call it stateless?
> 
> Could you rephrase that i couldn't understand

well looks interesting
im just thinking that eth foundation must have considered this already? maybe not because it (PoI) is new?

also i know that the eth foundation will do stateless proofing in the coming years after PoS merge, which might be similar to PoI

**Comment by JushBJJ** [comment]
> > > there must be more too it?
> > > or the eth devs had also chosen PoI,
> > > actually maybe they will have PoI also, i think they call it stateless?
> > 
> > Could you rephrase that i couldn't understand
> 
> well looks interesting
> im just thinking that eth foundation must have considered this already? maybe not because it (PoI) is new?
> 
> also i know that the eth foundation will do stateless proofing in the coming years after PoS merge, which might be similar to PoI

No one has tried PoI so why not try it now? Preferrably in a testnet first.

**Comment by simonmadsen** [comment]
might be a good idea.
maybe PoS first and then PoI after? built on top like i think eth will

**Comment by JushBJJ** [comment]
> might be a good idea.
> maybe PoS first and then PoI after? built on top like i think eth will

this isn't about ethereum

**Comment by simonmadsen** [comment]
> > might be a good idea.
> > maybe PoS first and then PoI after? built on top like i think eth will
> 
> this isn't about ethereum

i know, just referencing other peoples work
both of our points are valid

**Comment by JushBJJ** [comment]
> might be a good idea.
> maybe PoS first and then PoI after? built on top like i think eth will

I think its better to test out an algorithm that can potentionally disrupt PoW and PoS first, we know the mistakes of both algorithms so we learn from them and fix those mistakes. There is no point on testing PoS because we already know the outcome along with its pros and cons.

If we don't innovate fast enough then eventually we would be overtaken by something that is 10x better.

**Comment by rnicoll** [comment]
I think "own by the same provider as the server" is the issue here, this is intended for scenarios where you have imbalance across services run by a single entity, and want to slow access to one by having them interact with others first. It's a lot less clear if there's any way this could map to a network where thousands/tens of thousands of individuals run the infrastructure.

**Comment by JushBJJ** [comment]
<h3>Main Idea</h3>

>I think "own by the same provider as the server" is the issue here

>It's a lot less clear if there's any way this could map to a network where thousands/tens of thousands of individuals run the infrastructure.

Perhaps we could replace `server` with the set N of n nodes that isn't by the same provider. Meaning, we can first do the tour through n nodes given to lets say n1, n2, n3, and so on, these nodes could be lets say from different individuals in Australia, Europe, and Africa.

<h3>Optional Addition</h3>
Then to prove to the server, the user must prove to the n amount of nodes in the given set N. Moreover, the data included when the user tries to prove contains the nodes it will soon or is already proven to and will be stored for verification later. Each node except the first node will broadcast its role as the n'th node that the user has tried to prove. Of course this may not affect future nodes before the user interacts to them.

So the nodes before the current n'th node will know the current n'th node's existence. After that, the last node will broadcast to the first node of it's role, since the first node has a checklist of all the nodes that signaled it's existence, the first node will then generate the new block.

**Comment by patricklodder** [comment]
Hi! Do you have any code?

**Comment by JushBJJ** [comment]
> Hi! Do you have any code?

Unfortunately, the closest I could give you is the pseudo code from the paper which is located below the references. And functions used in the pictures below are described in the paper.

![image](https://user-images.githubusercontent.com/36951064/119536067-03395880-bdcc-11eb-9459-0188027b94f0.png)
![image](https://user-images.githubusercontent.com/36951064/119536100-0af8fd00-bdcc-11eb-94b0-54e9057dc064.png)
![image](https://user-images.githubusercontent.com/36951064/119536146-177d5580-bdcc-11eb-93e1-550e07bac84f.png)
![image](https://user-images.githubusercontent.com/36951064/119536182-219f5400-bdcc-11eb-8ee2-1723f4f6ade3.png)

**Comment by JushBJJ** [comment]
Should perhaps a pull request draft be created and start from there?

**Comment by patricklodder** [comment]
> Should perhaps a pull request draft be created and start from there?

Honestly, what I'd recommend you is to implement this on a separate repository in a proof of concept for. If you really intend to target Dogecoin, make it a regtest network integration, so that we can run it and see how it would integrate, and I'd like to discuss some ideas you would need to have about how you would convince miners that have been keeping Dogecoin secure & alive for almost 7 years now, to accept this, on your repository, not this one.

When that's there, a PR would be easier to attain, as it would be a waste of time to try and PR something that will be impossible to merge.

Hope that sounds reasonable? In the meantime we can talk about this on Reddit! https://www.reddit.com/r/dogecoindev

**Comment by JushBJJ** [comment]
> > Should perhaps a pull request draft be created and start from there?
> 
> Honestly, what I'd recommend you is to implement this on a separate repository in a proof of concept for. If you really intend to target Dogecoin, make it a regtest network integration, so that we can run it and see how it would integrate, and I'd like to discuss some ideas you would need to have about how you would convince miners that have been keeping Dogecoin secure & alive for almost 7 years now, to accept this, on your repository, not this one.
> 
> When that's there, a PR would be easier to attain, as it would be a waste of time to try and PR something that will be impossible to merge.
> 
> Hope that sounds reasonable? In the meantime we can talk about this on Reddit! https://www.reddit.com/r/dogecoindev

Absolutely! I was actually planning to implement the algorithm first in a separate repository but thought bringing it up here as an idea would be great.

Will get back to you when there are updates. 😄
Reddit Thread: https://www.reddit.com/r/dogecoindev/comments/nkvw4u/idea_switch_to_proof_of_interaction_an_algorithm/

**Comment by simonmadsen** [comment]
lol would be amazing to skip doge to the elite
